tests/student_comment_test.php

```diff
@@ -0,0 +1,206 @@
+
+/**
+ * Test student comments
+ *
+ * @copyright Davo Smith <moodle@davosmith.co.uk>
+ * @package   mod_checklist
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace mod_checklist;
+
+use context_module;
+use core\log\manager;
+use logstore_standard\log\store;
+use mod_checklist\local\checklist_comment_student;
+use mod_checklist\local\checklist_item;
+use mod_checklist_external;
+use stdClass;
+
+defined('MOODLE_INTERNAL') || die();
+
+/**
+ * Class student_comment_test
+ * @package mod_checklist
+ */
+class student_comment_test extends \advanced_testcase
+{
+    /** @var stdClass The student object. */
+    protected $student;
+
+    /** @var stdClass The course module object. */
+    protected $cm;
+
+    /** @var store The log manager object. */
+    protected $store;
+
+    /** @var checklist_item[] checklist items */
+    protected $items;
+    /**
+     * Set up steps
+     */
+    public function setUp(): void {
+        global $CFG, $USER, $DB;
+        $this->resetAfterTest();
+        // Create test checklist with a couple items.
+        require_once("$CFG->dirroot/mod/checklist/externallib.php");
+        $gen = self::getDataGenerator();
+        /** @var \mod_checklist_generator $cgen */
+        $cgen = $gen->get_plugin_generator('mod_checklist');
+
+        $c1 = $gen->create_course(['startdate' => strtotime('2019-04-10T12:00:00Z')]);
+        $chk = $cgen->create_instance(['course' => $c1->id]);
+
+        // Create a student who will add data to these checklists.
+        $this->student = $gen->create_user();
+        $studentrole = $DB->get_record('role', ['shortname' => 'student']);
+        $gen->enrol_user($this->student->id, $c1->id, $studentrole->id);
+        $this->setUser($this->student);
+
+        $iteminfos = [
+            (object)[
+                'displaytext' => 'Item 1',
+                'duetime' => strtotime('2019-05-14T12:00:00Z'),
+            ],
+            (object)[
+                'displaytext' => 'Item 2',
+            ],
+        ];
+        $position = 1;
+        foreach ($iteminfos as $iteminfo) {
+            $item = new \mod_checklist\local\checklist_item();
+            $item->checklist = $chk->id;
+            $item->userid = 0;
+            $item->displaytext = $iteminfo->displaytext;
+            $item->position = $position++;
+            $item->duetime = $iteminfo->duetime ?? null;
+            $checklistitemid = $item->insert();
+            $this->items[] = $item;
+            // Create student comments alongside the other items.
+            checklist_comment_student::update_or_create_student_comment($checklistitemid, 'testcomment' . $position, false);
+        }
+
+        $this->cm = get_coursemodule_from_instance('checklist', $chk->id, $c1->id);
+        $context = context_module::instance($this->cm->id);
+        $this->assertTrue(has_capability('mod/checklist:updateown', $context));
```

Comment:
I don't think there is any great benefit in testing the Moodle capability code - best to just focus on testing the code for this plugin (unless there is some subtle reason why you think has_capability() needs checking here?

Reply:
True, I'll remove this